### ISSUE TITLE
[Website] Website deployment workflow (`deploy.yml`) is failing due to Node.js 18 version bump in`ubuntu-latest` GitHub Actions runner image and Webpack usage of `md4` hashing algorithm

### ISSUE THREAD

**Issue by kevingurney** [apache/arrow#34379]
### Describe the bug, including details regarding any error messages, version, and platform.

[A few weeks ago](https://github.com/apache/arrow-site/actions/runs/4263179843/jobs/7419585498), the `apache-arrow-site` deployment workflow (`.github/workflows/deploy.yml`) started failing with the following output:

```shell
.
.
.
npm ci
npm WARN deprecated popper.js@1.16.1: You can find the new Popper v2 at @popperjs/core, this package is dedicated to the legacy v1

added [13](https://github.com/apache/arrow-site/actions/runs/4257427943/jobs/7407542742#step:9:14)2 packages, and audited 133 packages in 2s

17 packages are looking for funding
  run `npm fund` for details

found 0 vulnerabilities
npx webpack --mode=production
rm -f javascript/main.js
node:internal/crypto/hash:71
  this[kHandle] = new _Hash(algorithm, xofLen);
                  ^

Error: error:0308010C:digital envelope routines::unsupported
    at new Hash (node:internal/crypto/hash:71:19)
    at Object.createHash (node:crypto:133:10)
    at BulkUpdateDecorator.hashFactory (/home/runner/work/arrow-site/arrow-site/node_modules/webpack/lib/util/createHash.js:[14](https://github.com/apache/arrow-site/actions/runs/4257427943/jobs/7407542742#step:9:15)4:18)
    at BulkUpdateDecorator.update (/home/runner/work/arrow-site/arrow-site/node_modules/webpack/lib/util/createHash.js:46:50)
    at RawSource.updateHash (/home/runner/work/arrow-site/arrow-site/node_modules/webpack-sources/lib/RawSource.js:64:8)
    at NormalModule._initBuildHash (/home/runner/work/arrow-site/arrow-site/node_modules/webpack/lib/NormalModule.js:753:[17](https://github.com/apache/arrow-site/actions/runs/4257427943/jobs/7407542742#step:9:18))
    at handleParseResult (/home/runner/work/arrow-site/arrow-site/node_modules/webpack/lib/NormalModule.js:817:10)
    at /home/runner/work/arrow-site/arrow-site/node_modules/webpack/lib/NormalModule.js:908:4
    at processResult (/home/runner/work/arrow-site/arrow-site/node_modules/webpack/lib/NormalModule.js:640:11)
    at /home/runner/work/arrow-site/arrow-site/node_modules/webpack/lib/NormalModule.js:692:5 {
  opensslErrorStack: [ 'error:03000086:digital envelope routines::initialization error' ],
  library: 'digital envelope routines',
  reason: 'unsupported',
  code: 'ERR_OSSL_EVP_UNSUPPORTED'
}

Node.js v[18](https://github.com/apache/arrow-site/actions/runs/4257427943/jobs/7407542742#step:9:19).14.1
rake aborted!
Command failed with status (1): [npx webpack --mode=production...]
/home/runner/work/arrow-site/arrow-site/Rakefile:37:in `block in <top (required)>'
/home/runner/work/arrow-site/arrow-site/vendor/bundle/ruby/3.0.0/gems/rake-13.0.6/exe/rake:27:in `<top (required)>'
/opt/hostedtoolcache/Ruby/3.0.2/x64/bin/bundle:[23](https://github.com/apache/arrow-site/actions/runs/4257427943/jobs/7407542742#step:9:24):in `load'
/opt/hostedtoolcache/Ruby/3.0.2/x64/bin/bundle:23:in `<main>'
Tasks: TOP => generate => javascript/main.js
(See full trace by running task with --trace)
Error: Process completed with exit code 1.
```
This appears to be related to the use of Webpack by `apache-arrow-site` and the following issues:

1. https://github.com/webpack/webpack/issues/14532#issuecomment-1434095919
2. https://github.com/webpack/webpack/issues/13572
3. https://github.com/webpack/webpack/pull/14306

My high level understanding is that in Node 18 (the build output above shows Node.js `v18.14.1` is being used), the `md4` hashing algorithm is deprecated (more specifically, it seems that [Node 18 uses OpenSSL 3.0](https://github.com/nodejs/TSC/blob/main/OpenSSL-Strategy.md#nodejs-version-18x-est-apr-2022-eol-apr-2025), which has [deprecated `md4`](https://www.openssl.org/docs/man3.0/man3/MD4.html)) and the version of Webpack used by `apache/arrow-site` (`v5.21.2`) seems to default to using `md4`.

[Webpack v5.61.0](https://github.com/webpack/webpack/releases/tag/v5.61.0) added a WASM `md4` implementation as a fallback. However, the advice in https://github.com/webpack/webpack/issues/14532#issuecomment-1434095919 recommends setting `output.hasFunction` in the Webpack config to use an alternative hashing algorithm instead. Specifically, it recommends using `xxhash64` (which is planned to be the default hashing algorithm when Webpack 6 is released).

It seems that the [version of Node.js in the `ubuntu-latest` GitHub Actions runner image (used by `deploy.yml`) was bumped to v18 on Februrary 13, 2023](https://github.com/actions/runner-images/issues/7002). This would explain why this issue started appearing a few weeks ago.

## Workarounds

We could choose to [pin the version of Node.js used by the GitHub Actions runner to v16](https://github.com/actions/setup-node#usage) for the `actions/setup-node` action to work around this issue. Of course, this would mean we would be continuing to rely on an outdated version of Node.js, which doesn't seem ideal in the long term.

Perhaps, the more preferable option, would be to follow the advice in https://github.com/webpack/webpack/issues/14532#issuecomment-1434095919 and set `output.hasFunction` in the Webpack config to use an alternative hashing algorithm, like `xxhash64`.

### Component(s)

Website

**Comment by avantgardnerio** [comment]
FWIW, I'm in favor of both 2 & 3: we should control our upgrades and not have them unexpectedly change, but we should also follow a reasonable upgrade cycle and not use outdated cryptographic functions.

That being said, the arrow project in general lacks a surplus of front-end development skills, so it option 3 makes the most sense to me if we only have the resources for 1 of the above.

**Comment by kevingurney** [comment]
@avantgardnerio - thanks for sharing your thoughts.

I agree - 2 + 3 seems like the optimal solution. As you mentioned, if we are concerned with resourcing, starting with option 3 feels like a good first step which would unblock things in the short term. We could always choose to address 2 in a separate PR when resourcing permits.

**Comment by kevingurney** [comment]
take
